### PR TITLE
feat: Add pallet bags list

### PR DESCRIPTION
**Pull request by rdoria1**
The `runtime/src/bag_thresholds.rs` should be autogenerated (similar to the weight file) for our network.

generate bags script added with following command line

`cargo run -p node-runtime-generate-bags -- --node-url wss://testnet-rpc.zkverify.io runtime/bag_thresholds.rs`

### CODE REVIEW

**Comment by la10736** [comment]
I think you should also add here the `generate_bags`: in this case isn't relevant if this code run on reference machine or not because the result depends only by how our runtime is configured and the values used to generate it (`n_bags`, `total_issuance` and `minimum_balance`).

We can also chose to take `minimum_balance == existential_deposit` and remove it from the command line values and make the process less error prone by read it from runtime. Maybe for us could be also useful set a default for `n_bags` and provide an url to read the total issuance (where the default value point to testnet).

Just to summarize:
- copy the code in https://github.com/paritytech/polkadot-sdk/tree/master/substrate/utils/frame/generate-bags in our `utils` folder. 
- Adapt the `node-runtime` code to use our runtime 
- Get the default value for `minimum_balance` from the e`existential_deposit` 
- (Optional) read the total issuance from the given chain

File: utils/generate-bags/Cargo.toml
Comment (drgora) [inline utils/generate-bags/Cargo.toml:16]: Can we use the `workspace` versions here, as we do in the rest of the project?

File: utils/generate-bags/node-runtime/Cargo.toml
Comment (drgora) [inline utils/generate-bags/node-runtime/Cargo.toml:14]: I'd suggest to skip the version tag (which is implicit) when you pass a local path

File: utils/generate-bags/src/lib.rs
Comment (DanieleDiBenedetto) [inline utils/generate-bags/src/lib.rs:2]: Shouldn't we add copyright header for Parity here as well  ?

File: utils/generate-bags/node-runtime/src/main.rs
Comment (drgora) [inline utils/generate-bags/node-runtime/src/main.rs:63]: I believe it would be better to fetch the `existential_deposit` as a constant from the given RPC, similarly to what you do with the `total_issuance`, instead of taking it from the local code.

File: utils/generate-bags/src/lib.rs
Comment (DanieleDiBenedetto) [inline utils/generate-bags/src/lib.rs:128]: I assume yes, but please double check with John if this way of computing the bags thresholds is fine.

File: utils/generate-bags/node-runtime/src/main.rs
Comment (DanieleDiBenedetto) [inline utils/generate-bags/node-runtime/src/main.rs:28]: Let's probably agree with John how to set this value

**Review by drgora — CHANGES_REQUESTED** [review]
Thank you for adding the new pallet. The main thing that I believe we should do before merging, is to actually use the new pallet to keep track of the voters.
This should be done in the configuration of the Staking pallet. I also left few other comments that might be worth addressing.

**Comment by DanieleDiBenedetto** [comment]
> I think you should also add here the `generate_bags`: in this case isn't relevant if this code run on reference machine or not because the result depends only by how our runtime is configured and the values used to generate it (`n_bags`, `total_issuance` and `minimum_balance`).
> 
> We can also chose to take `minimum_balance == existential_deposit` and remove it from the command line values and make the process less error prone by read it from runtime. Maybe for us could be also useful set a default for `n_bags` and provide an url to read the total issuance (where the default value point to testnet).
> 
> Just to summarize:
> 
> * copy the code in https://github.com/paritytech/polkadot-sdk/tree/master/substrate/utils/frame/generate-bags in our `utils` folder.
> * Adapt the `node-runtime` code to use our runtime
> * Get the default value for `minimum_balance` from the e`existential_deposit`
> * (Optional) read the total issuance from the given chain

Not a big fan of tying this tool to an actual running/configured chain. I see Polkadot is exposing these values as input parameters for the CLI perhaps it's better to do in this way (so one can also do some local experiments if desired), or maybe encoding a logic s.t., if they are not specified, the code attempts to read it from the chain ?

File: runtime/src/lib.rs
Comment (DanieleDiBenedetto) [inline runtime/src/lib.rs:288]: Shouldn't this be taken from the election pallet @drgora ?

**Comment by DanieleDiBenedetto** [comment]
Also, let's not forget to run the benches for the staking pallet, now that we have integrated the bags list pallet
https://github.com/HorizenLabs/zkVerify/blob/main/scripts/run_all_benchmarks.sh#L109

File: utils/generate-bags/node-runtime/src/main.rs
Comment (rdoria1) [reply utils/generate-bags/node-runtime/src/main.rs:63]: Unfortunately I couldn't  retrieve it from API nor RPC

I found two "potential" ways to get the value, but none of them work (probably something is not implemented)

-> https://polkadot.js.org/docs/substrate/runtime#queryexistentialdepositcurrencyid-raw-u128 but tokens API is not implemented in the API

-> Following from integ tests in subxt https://github.com/paritytech/subxt/blob/9739a1682c5e81c7bfe539a08c3c1b0fb3e4798f/testing/integration-tests/src/full_client/metadata/validation.rs#L122
Same, doesn't work 

I'll keep it as it is for now.

File: utils/generate-bags/src/lib.rs
Comment (rdoria1) [reply utils/generate-bags/src/lib.rs:128]: asked John, waiting on reply

File: utils/generate-bags/node-runtime/src/main.rs
Comment (rdoria1) [reply utils/generate-bags/node-runtime/src/main.rs:28]: asked John, waiting on reply

File: utils/generate-bags/node-runtime/src/main.rs
Comment (drgora) [reply utils/generate-bags/node-runtime/src/main.rs]: Ok but this is weird, I would expect the latter to work on our testnet, the constants are accessible from polkadot-js, and there's also an example code that looks quite similar to yours: https://github.com/paritytech/subxt/blob/9739a1682c5e81c7bfe539a08c3c1b0fb3e4798f/subxt/examples/constants_static.rs

We should investigate this at some point.

File: runtime/src/lib.rs
Comment (drgora) [reply runtime/src/lib.rs:288]: Oh yes, good catch. `VoteWeight` would be better be taken from `sp-npos-elections`, which we already import indirectly.

File: utils/generate-bags/src/lib.rs
Comment (rdoria1) [reply utils/generate-bags/src/lib.rs:128]: this is fine, confirmed

File: utils/generate-bags/node-runtime/src/main.rs
Comment (rdoria1) [reply utils/generate-bags/node-runtime/src/main.rs:28]: 200 is fine for now, asked John

File: utils/generate-bags/node-runtime/src/main.rs
Comment (rdoria1) [reply utils/generate-bags/node-runtime/src/main.rs]: @drgora 
I'll put a ticket on the backlog

File: utils/generate-bags/src/lib.rs
Comment (rdoria1) [reply utils/generate-bags/src/lib.rs:2]: Added both as suggested by @la10736 (synced offline)

**Review by DanieleDiBenedetto — APPROVED** [review]

**Review by drgora — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
